#517 Redesign gardens index/list page

borracciaBlu opened the pull request:
Fixes #517 Redesign gardens index/list page

tygriffin commented on app/views/gardens/_plantings.html.haml:
```diff
@@ -0,0 +1,3 @@
+- if !plantings.empty?
```
Please use `unless plantings.empty?` or `if plantings.present?`. 

`empty?` and `present?` aren't exact inverses of each other, by the way. Please check this [SO question](http://stackoverflow.com/questions/885414/a-concise-explanation-of-nil-v-empty-v-blank-in-ruby-on-rails) if you're not familiar with the details.

tygriffin commented on app/views/gardens/index.html.haml:
```diff
@@ -22,43 +21,38 @@
+        
+        %td.status
+          - if garden.active 
+            %spam.label.label-success Active
```
Is this a typo? Should it be `span`?

borracciaBlu replied:
Hi @tygriffin yes it is, sorry

tygriffin commented on app/views/gardens/index.html.haml:
```diff
@@ -22,43 +21,38 @@
+          = render :partial => 'area' , locals: {:garden => garden} 
+            
         %td
           - if ! garden.location.blank?
```
Also, here: `if garden.location.present?` or `unless garden.location.blank?`

(Oh, I just realized you didn't write this line... but would be a nice refactor :) )

tygriffin commented on spec/views/gardens/_area.html.haml_spec.rb:
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
```
Unfortunately, view specs are deprecated! Sorry this wasn't more obvious. There's a PR for adding notice at the top of each spec so it will be more clear not to write these. Would you mind removing this spec and replacing it with a feature spec? 

You can probably integrate a test for these in the existing garden spec: `spec/features/gardens_spec.rb`. And probably no need to check every case - just probably no area and one other.

Additionally, we are using the `expect` syntax. So please replace lines like `response.should match(/m\n<sup>2<\/sup>/i)` with something like: `expect(response).to match(/m\n<sup>2<\/sup>/i)`

tygriffin commented on app/views/gardens/_area.html.haml:
```diff
@@ -0,0 +1,15 @@
+    - elsif garden.area_unit === 'hectare'
+      %span{:title => 'hectare'} HA
+    - elsif garden.area_unit === 'acre'
+      %span{:title => 'acre'} A
```
I wonder if this would be more re-useable as a view helper. Also, instead of a long conditional, I might make a hash that acts as dictionary for these abbreviations, like this:

```
area_abbreviations = {
  'square meter' => 'm<sup>2</sup>',
  # etc
}
```
That way you can just look up the abbreviation with `#{area_abbreviations[garden.area_unit]}`

Just my personal opinion though :)

borracciaBlu replied:
Hi @tygriffin I like it too :P
Just some questions..
Where should I put the code? Should i generate a new helper or a variable in the partial is good enough?

tygriffin replied:
You can make a new file in `app/helpers`. You can return an html string from the helper and just call `html_safe` on it to make sure all your brackets aren't escaped.

tygriffin commented on app/views/gardens/index.html.haml:
```diff
@@ -22,43 +21,38 @@
     - @gardens.each do |garden|
       %tr
+        
+        %td.status
```
Any special reason why there's a class of `status` here? I think this and the classes `name` and `area` below might be a little too generic. We need to have conversation generally about how to organize CSS within Growstuff, so until then maybe better to remove these.

borracciaBlu replied:
Hi @tygriffin yes the reason is in the css. 
I need to target .label-success just in this page and to do that i'm using .status .label-success.
I didn't use .active because it was changing the color of the column

tygriffin replied:
Oh, I see what you mean. The trouble is that all this CSS is in a global namespace so it's pretty easy to create conflicts with generic names like "status" or "area".

Can we change "status" to "garden-status" and remove the "name" and "area" classes?

borracciaBlu replied:
yes, it would be nice to put a unique id in the css in each page, like #garden_index.
In that way after in less we could use :
#garden_index {
   .garden-status {
   }
}

definetly even the organisation of the less file is far away to be perfect.

tygriffin commented:
Nice! Thanks for the PR. I left some inline notes but looks like you've addressed half of them as fast as I could write them :)

For these design related PR's it might be nice to include a screenshot so we can get a sense of what you've changed. (I need to take my own advice and do this for one of my PR's).

tygriffin commented:
By the way, when you make a PR please reference the issue you're addressing like this:

"Fixes #517"

That way, you're PR will be connected to that issue. When the PR is merged, the issue will be automatically closed. I've already edited this PR though.

borracciaBlu commented:
Where should I put the screenshot?

tygriffin commented:
You can just paste the pic right here in a comment to this PR.

tygriffin commented on app/views/gardens/_plantings.html.haml:
```diff
@@ -0,0 +1,3 @@
+- unless plantings.empty?
+  - plantings.each do |p|
+    = link_to p.crop.name, p, :class => "label label-info", :title => "#{p.quantity} #{p.crop.name} #{p.planted_at}"
```
I see why you've put the crop info in a title here so the plantings column won't be so cluttered. This is not a bad idea, but one alternative is a [Bootstrap tooltip](http://getbootstrap.com/javascript/#tooltips). I think they're slightly more responsive than titles. 

Before changing anything, it's probably a good idea to get input from someone who uses Growstuff on a regular basis, like @Skud or @maco

borracciaBlu commented:
![screenshot](https://cloud.githubusercontent.com/assets/2061731/6205500/071365d4-b5c4-11e4-90d3-246077b4bdfb.png)
That's how it looks like.
Main changes:
* moved active | inactive in first column and added label + colors
* removed description
* reduced information about area using ISO format
* reduced info about planting using label and title

Skud commented:
OK, so, this is a pretty major change to look and feel for the site. 
It's using visual styles and colours that aren't used elsewhere. I would 
very much rather *not* do this at this time!  If we are going to do 
something like this, we should do it across the whole site.

I think we should also consider 
https://github.com/Growstuff/growstuff/issues/570 "consistent visual 
language for thumbnails" -- the desire there is to have a more 
thumbnail-based approach to displaying things with photos and some info, 
rather than the tabular layout (or having them both as an option as per 
the wireframes at 
http://talk.growstuff.org/t/aspirational-wireframes-for-the-crop-detail-page/204)

But the big picture is: when I come to look at a list of my gardens, 
what do I want to see?

When I'm looking at my own gardens, as a gardener my needs are:

* show the active gardens most prominently (inactive ones can be almost 
hidden; I've probably moved house or something, and don't use them any more)
* a quick overview of what state each garden's in
** what does it look like (photo, in other words)
** what is currently planted in it? (don't count finished plantings)
** maybe a note about when I last planted something in the garden, to 
prompt me if I've forgotten to update it recently
* quick action links to do stuff with it:
** plant something in it
** edit details
** add a photo

When I'm looking at other people's gardens, there's no real action I can 
take with them, so I think I'll mostly be in "ooh shiny" mode, wanting 
to browse around for interesting and active gardens. So again I'd like 
to see a visual overview of the gardens, and I think it would be good to 
get a sense of the level of activity in a given garden, maybe sort by 
recently active, or something like that.

Sorry to be coming in at this late stage with such a high level 
comment.  Ideally for these "visual design" stories we should do some 
wireframes and get feedback on the general ideas early on, before 
touching the code.  I can highly recommend http://balsamiq.com for 
wireframing -- they have a free browser based version or a paid desktop one.

A.


On 16/02/2015 10:11 am, borracciaBlu wrote:
>
> screenshot 
> <https://cloud.githubusercontent.com/assets/2061731/6205500/071365d4-b5c4-11e4-90d3-246077b4bdfb.png>
> That's how it looks like.
> Main changes:
>
>   * moved active | inactive in first column and added label + colors
>   * removed description
>   * reduced information about area using ISO format
>   * reduced info about planting using label and title
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Growstuff/growstuff/pull/709#issuecomment-74442845>.
>

-- 
Alex "Skud" Bayley
skud@growstuff.org
http://growstuff.org/

borracciaBlu commented:
Sorry I was committing..

Skud commented:
No, it's great that you took a shot at it :)  The issue in Github was 
very badly described -- we should at least add a "please open a 
discussion on this subject before starting to code" to the description 
of these "needs design" tasks!  We are growing very fast and some of our 
cultural norms need to change as this happens. Sorry you got caught up 
in it.

Are you interested in looking at the thumbnail stuff?  You seem to have 
a pretty good eye for frontend/design stuff, so I would love to hear 
your ideas on it.  Maybe I could mock something up for discussion and 
ping you about it? Or feel free to do so yourself if you like.

A.

On 16/02/2015 10:44 am, borracciaBlu wrote:
>
> Sorry I was committing..
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Growstuff/growstuff/pull/709#issuecomment-74444204>.
>

-- 
Alex "Skud" Bayley
skud@growstuff.org
http://growstuff.org/

borracciaBlu commented:
Yes It would be great!
Definetly we should open a discussion about that and the structure of the less files.

Here I think we have two main steps:
1 - Restyle the current page without change the informations and the actions
2 - Start to re-think the current informations in the thumbnails way for this I think we can create a new discussion 

About the point 1:
Ok I think that as first step I can remove the classes responsables for the visual changes.
In that way we are not introducing any new visual element and we are delivering the current informations in a better shape.

About the point 2:
Create a new dicussion called maybe Garden Card (Why card? [Cards in web design](http://webdesignledger.com/trends-2/cardbaseddesign).
Insert in the discussion all the info you suggested before.
Start to make mockup and discuss.

What do you think?

![screenshot2](https://cloud.githubusercontent.com/assets/2061731/6205830/e25afb06-b5cf-11e4-8204-d4789f8685f0.png)
That's how it looks like without the labels classes

Skud commented:
1. Yes, "card based design" is a great term, and we should update #570 to reflect that!  100% agree, let's have a discussion on [Ideas](http://talk.growstuff.org/c/idea) about it.

2. Just looking at the (reduced) changes to the garden page, yes, I think you are on the right track here!  Some notes:

* I think we should use "acres" and "hectares" spelled out.  The words aren't too long and the abbreviations are little-known ("acres" doesn't have a standard one as far as I can tell, and "hectares" is "ha" according to SI).  I agree, though, that m<sup>2</sup> is nicer than "square metres".  OTOH, "sqft" is way more common than "ft<sup>2</sup>" in my experience.  Overall my feeling is "don't abbreviate" to be honest.
* The crops list should be comma separated. 
* What happens if there are multiple of the same crop, eg "zucchini, zucchini, zucchini".  I see there was some discussion of title vs bootstrap tooltips; I think we should use the bootstrap options where available.
* Please make sure that only current (not finished) plantings are listed.
* I'm not sure about the ordering of Active/Inactive being the first column, especially when you're viewing everyone's gardens.  But I also don't have very strong feelings, so I guess it's ok either way.

borracciaBlu commented:
- I understand your point but if we spell acres and hectares we still have the issue of the plural and my perception about hectares is that is too long.
What do you think if we spell just acres and for hectares we use ha instead of HA.
Consider that even in this case we have the title attribute in.
About the sqft yep that's fine np. 
- the crops list now is comma separated 
- about the tooltip ok np
- ok for the "not finished", i will work on that tomorrow

cesy commented:
acres and ha sounds like a reasonable compromise to me, as none of the options are ideal. @borracciaBlu, are you still keen to finish this? Sorry for the long delay.

borracciaBlu commented:
No worries. 

As far as I know it was completed. Btw if you want   I'll can change it.

cesy commented on app/assets/javascripts/gardens.js.erb:
```diff
@@ -0,0 +1,3 @@
+$(function () {
+  $('[data-toggle="tooltip"]').tooltip()
+})
```
Can you add a blank line here to make Github happy, please?

cesy commented on app/assets/stylesheets/gardens.css.less:
```diff
@@ -0,0 +1,4 @@
+.buttons-bar {
+	margin-top: 32px;
+	margin-bottom: 0px;
+}
```
Same here.

cesy commented on app/assets/stylesheets/overrides.css.less:
```diff
@@ -22,10 +22,10 @@ body {
 
 // Font Awesome
-//@import "fontawesome/font-awesome";
+// @import "fontawesome/font-awesome";
```
What's the reason for adding the space here?

borracciaBlu replied:
Ahahaha well ... just aesthetic..

cesy commented on app/helpers/gardens_helper.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+  end
+
+end
```
Blank line, please.

cesy commented on app/helpers/gardens_helper.rb:
```diff
@@ -0,0 +1,18 @@
+    area_abbreviations = {
+      'square metre' => 'm<sup>2</sup>',
+      'square foot'  => 'sqft',
+      'hectare'      => 'ha',
```
Great, so we can change this easily in the future if needed.

cesy commented on spec/helpers/gardens_helper_spec.rb:
```diff
@@ -0,0 +1,27 @@
+    end
+  end
+
+end
```
I can't see the test that uses this new functionality - am I missing something?

borracciaBlu replied:
i wrote the test before write the code (TDD). If you like i can remove it btw.

cesy replied:
TDD is good, and tests are good, but can you add a feature test to cover this, please?

borracciaBlu replied:
@cesy i'm sorry but I'm a little out of time in this period so feel free to change anything you think is good. If you need i can close the pull request and you can pull from my repo make the changes you need and pull it. Let me know

pozorvlak commented:
#768 also addresses this issue in a card-based manner. If we merge #768 we should close this. @borracciaBlu, I'm really sorry this has dragged on for so long :-(

pozorvlak commented:
Superseded by #768 - @borracciaBlu, thank you for your hard work, and I'm sorry we didn't end up incorporating it.
